tfjs-backend-webgpu/src/softmax_webgpu.ts

```diff
@@ -16,36 +16,89 @@
  */
 
 import {getMainHeaderString as main, WebGPUProgram} from './webgpu_program';
-import {computeDispatch, flatDispatchLayout} from './webgpu_util';
+import {flatDispatchLayout} from './webgpu_util';
 
-// This kernel in fact is a combination of sub, exp.
 export class SoftmaxProgram implements WebGPUProgram {
-  variableNames = ['a', 'b'];
+  variableNames = ['logits'];
   outputShape: number[];
   shaderKey: string;
   dispatchLayout: {x: number[]};
   dispatch: [number, number, number];
-  workgroupSize: [number, number, number] = [64, 1, 1];
-  size = true;
+  workgroupSize: [number, number, number];
 
   constructor(outputShape: number[]) {
-    this.outputShape = outputShape;
+    this.outputShape = outputShape;  // [rows, cols]
     this.dispatchLayout = flatDispatchLayout(this.outputShape);
-    this.dispatch = computeDispatch(
-        this.dispatchLayout, this.outputShape, this.workgroupSize);
+    this.dispatch = [this.outputShape[0], 1, 1];
+    if (this.outputShape[1] >= 4096) {
+      this.workgroupSize = [256, 1, 1];
+    } else {
+      this.workgroupSize = [64, 1, 1];
+    }
     this.shaderKey = 'softmax';
   }
 
   getUserCode(): string {
     const userCode = `
-      ${main('index')} {
-        if (index < uniforms.size) {
-          let a = getAByOutputIndex(index);
-          let b = getBByOutputIndex(index);
-          let value = a - b;
-          setOutputAtIndex(index, exp(value));
+    var<workgroup> buf : array<f32, ${this.workgroupSize[0]}>;
+    var<workgroup> rowMaxShared : f32;
+    var<workgroup> rowSumShared : f32;
+    const blockSize = ${this.workgroupSize[0]};
+    ${main('index')} {
+      let row = index / blockSize;
+      let tid = i32(localId.x);
+      let cols = uniforms.outShape[1];
+
+      var threadMax = -1.0 / 1e-20;
+      for (var col = tid; col < cols; col += blockSize) {
+        let value = getLogits(row, col);
+        threadMax = max(threadMax, value);
+      }
+      if (tid < cols) {
+        buf[tid] = threadMax;
+      }
+      workgroupBarrier();
+
+      let reduceSize = min(cols, blockSize);
+      for (var currSize = reduceSize >> 1;  currSize > 0; currSize = currSize >> 1) {
+        if (tid < currSize) {
+          buf[tid] = max(buf[tid], buf[tid + currSize]);
         }
+        workgroupBarrier();
+      }
+
+      if (tid == 0) {
+        rowMaxShared = max(buf[0], buf[reduceSize - 1]);
+      }
+      workgroupBarrier();
+
+      var threadSum = 0.0;
+      for (var col = tid; col < cols; col += blockSize) {
+        let subExp = exp(getLogits(row, col) - rowMaxShared);
+        threadSum += subExp;
+      }
+      if (tid < cols) {
```

Comment:
Because we will sum all elements of `buf` below, if cols is less than blockSize, will there be an error? I think the `if` condition could be removed here.